### ISSUE TITLE
Server Side Rendering - Static Site Generation

### ISSUE THREAD

**Issue by tomchentw** [tomchentw/reacthtmlpack#5]
## Goals

I'm much more interested in I'm targeting:

* Single `index.html.js` file
* React-Router based application
* Each routes mapped to different pages
* Each pages are the combination of `index.html.js` and React-Router rendered result

By __Single `index.html.js` file__, that means
* Each pages will have the same set of css/js files, it's the application's responsibility to dynamically dispatch different styles for different pages
* Each page may contain different title/meta tags supported by react-helmet, but again the dispatch logic should exist in react-router application.

This allows us to build several static hosted pages with one react-router backed application. For example, company website, blog, personal website, library demo pages, library documentation pages……

#### index.html.js

```js
import {
  default as React,
} from "react";

import {
  WebpackScriptEntry,
  WebpackStyleEntry,
} from "reacthtmlpack/lib/entry";

export default (
  <html>
    <head>
      <title>reacthtmlpack | Static Site Generation</title>
      <meta name="viewport" content="width=device-width, initial-scale=1"/>
      <meta charSet="UTF-8" />
      <WebpackStyleEntry
        chunkName="main"
        chunkFilepath="../scripts/main.js"
        configFilepath="../StaticSiteGeneration.webpackConfig.js"
      />
    </head>
    <body>
      <div id="react-container" />
      <WebpackScriptEntry
        chunkName="main"
        chunkFilepath="../scripts/main.js"
        configFilepath="../StaticSiteGeneration.webpackConfig.js"
      />
    </body>
  </html>
);
```

#### routes.js

```js
import {
  default as React,
} from "react";

import {
  Route,
} from "react-router";

export default (
  <Route path="/" component={App}>
    <Route path="home" component={Home}/>
    <Route path="contact" component={Contact}>
    <Route path="*" component={NoMatch}/>
  </Route>
);
```

### client.js

```js
import React from "react";
import { render } from "react-dom";
import { Router, browserHistory } from "react-router";

import {
  default as routes,
} from "./routes";

render((
  <Router
    history={browserHistory}
    routes={routes}
  />
), document.getElementById(`react-container`));
```

### server.js

https://github.com/rackt/react-router/blob/latest/docs/guides/advanced/ServerRendering.md

```js
import React from "react";
import { renderToString } from "react-dom/server";
import { match, RouterContext } from "react-router";

import {
  default as routes,
} from "./routes";

const matchArgumentsList = [
  { location: `/` },
  { location: `/home` },
  { location: `/contact` },
];

export default () => {
  // Return Array<Promise<String>>
  return Promise.all(matchArgumentsList.map(({location}) => {
    // match accepts callback, so we need to wrap it in a Promise
    return new Promise((resolve, reject) => {
      match({ routes, location }, (error, redirectLocation, renderProps) => {
        if (renderProps) {
          resolve(
            // The dom tree that can be used by client-side rendering
            renderToString(<RouterContext {...renderProps} />)
          );
        } else {
          reject(error || redirectLocation || 'Not found');
        }
      });
    });
  }));
};
```

### Expected results

```log
public/index.html
public/home.html
public/contact.html
public/404.html
public/assets/main.js
public/assets/main.css
```

## Related libraries

* http://jxnblk.com/writing/posts/static-site-generation-with-react-and-webpack/
* https://github.com/markdalgleish/static-site-generator-webpack-plugin
* https://github.com/nfl/react-helmet

**Comment by tomchentw** [comment]
https://github.com/markdalgleish/static-site-generator-webpack-plugin

**Comment by tomchentw** [comment]
https://github.com/nfl/react-helmet

**Comment by tomchentw** [comment]
Closed by #20
